mvc/models/model-binding.rst

```diff
@@ -1,8 +1,89 @@
+
+Since the routing template looks like this, ``{controller=Home}/{action=Index}/{id?}``, ``movies/edit/2`` routes to the ``Movies`` controller, and its ``Edit`` action method. It also accepts an optional parameter called ``id``. This is due to MVC's use of conventions. So the code for the action method should look something like this: 
+
+.. literalinclude:: model-binding/sample/src/MVCMovie/Controllers/MoviesController.cs
+   :language: c#
+   :lines: 61-76  
+   
+.. note:: The strings in the URL route are not case sensitive.
+
+MVC checks for model binding attributes applied to the action method's parameters. If it finds a parameter attribute it will use the data source specified by the attribute. Otherwise, for any property that has a string converter, it looks for a value with a matching name. In the above example the only matching item is an integer named `id`, so then the value provider associates its values with action method parameters. Below is a list of the data sources in the order that model binding looks through them:
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method. (including jQuery POST requests).
+#. ``Route values``: The set of route values provided by `routing <https://docs.asp.net/projects/mvc/en/latest/controllers/routing.html>`_. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked for a key named ``id`` and there is nothing named ``id`` in the form values, it moved on to the route values looking for that key. In our example, it's a match. Binding happens, and the value is converted to the integer 2. The same request using Edit(string id) would convert to the string "2". 
+
+So far the example uses primitive types. If the action method's parameter were a complex type such as the ``Movie`` type, which contains both primitive and complex types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types looking for matches. Model binding looks for the pattern `parameter_name.property_name` to perform binding. If it doesn't find matching objects, it will attempt to bind using  just the property name. For complex types such as ``Collection`` types model binding looks for matches to `parameter_name[index]` or just `[index]`. Model binding treats  `Dictionary`` types similarly, asking for `parameter_name[key]` or just `[key]`, as long as they keys are simple types. 
+
+In order for binding to happen, members of complex model types must be public, writable properties containing a default public constructor. Public fields are not bound. If properties are arrays or IEnumerable types, the array properties or types in the IEnumerable must be writable. Model binding handles some objects differently than most. ``ICollection<T>`` is supported for read only properties.  
```

Comment:
> Collection properties that have an `Add` method can be read-only.

@danroth27 you might want to reword this sentence because the system does _not_ search for any olde `Add()` method, model binding special cases read-only properties of types that implement `ICollection<T>`.